pkg/util/downloader/resource_test.go

```diff
@@ -0,0 +1,96 @@
+package downloader_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/devstream-io/devstream/pkg/util/downloader"
+	"github.com/onsi/gomega/ghttp"
+)
+
+var _ = Describe("ResourceClient struct", func() {
+	var (
+		source string
+		c      *downloader.ResourceClient
+	)
+	BeforeEach(func() {
+		c = &downloader.ResourceClient{}
+	})
+	Context("GetWithGoGetter method", func() {
+		When("source is local files or directory", func() {
+			BeforeEach(func() {
+				source = GinkgoT().TempDir()
+				c.Source = source
+			})
+			It("should return local source path directly", func() {
+				dstPath, err := c.GetWithGoGetter()
+				Expect(err).Error().ShouldNot(HaveOccurred())
+				Expect(dstPath).Should(Equal(source))
+			})
+		})
+		When("download resource from web", func() {
+			var (
+				s       *ghttp.Server
+				reqPath string
+			)
+			BeforeEach(func() {
+				s = ghttp.NewServer()
+			})
+			When("resource return error", func() {
+				BeforeEach(func() {
+					reqPath = path.Join(s.URL(), "not_exist_resource")
+					c.Source = reqPath
+				})
+				It("should return err", func() {
+					_, err := c.GetWithGoGetter()
+					Expect(err).Error().Should(HaveOccurred())
+					Expect(err.Error()).Should(ContainSubstring("http: no Host in request URL"))
+				})
+			})
+			When("resource return normal", func() {
+				var (
+					returnContent string
+				)
+				BeforeEach(func() {
+					returnContent = "test getter"
+					pathName := "/exist_resource"
+					reqPath = fmt.Sprintf("%s%s", s.URL(), pathName)
+
+					s.RouteToHandler("HEAD", pathName, func(w http.ResponseWriter, r *http.Request) {
+						fmt.Fprint(w, "ok")
+					})
+					s.RouteToHandler("GET", pathName, ghttp.CombineHandlers(
+						ghttp.VerifyRequest("GET", pathName),
+						ghttp.RespondWith(http.StatusOK, returnContent),
+					))
+					c.Source = reqPath
+				})
+				When("destination is not setted", func() {
+					It("should create temp dir and download file", func() {
+						dstPath, err := c.GetWithGoGetter()
+						Expect(err).Error().ShouldNot(HaveOccurred())
+						files, err := ioutil.ReadDir(dstPath)
```

Comment:
// ReadDir reads the directory named by dirname and returns
// a list of fs.FileInfo for the directory's contents,
// sorted by filename. If an error occurs reading the directory,
// ReadDir returns no directory entries along with the error.
//
// Deprecated: As of Go 1.16, os.ReadDir is a more efficient and correct choice:
// it returns a list of fs.DirEntry instead of fs.FileInfo,
// and it returns partial results in the case of an error
// midway through reading a directory.
//
// If you must continue obtaining a list of fs.FileInfo, you still can:
//
//	entries, err := os.ReadDir(dirname)
//	if err != nil { ... }
//	infos := make([]fs.FileInfo, 0, len(entries))
//	for _, entry := range entries {
//		info, err := entry.Info()
//		if err != nil { ... }
//		infos = append(infos, info)
//	}